### ISSUE TITLE
[BUG]: It is not possible to make connection between different (public) networks 

### ISSUE THREAD

**Issue by emrahtoy** [Unity-Technologies/UnityRenderStreaming#907]
### Package version

3.1.0-exp.6

### Environment

```markdown
* OS: Windows 10 and Windows 11, Linux Ubuntu 22.04 Docker env.
* Unity version: 2021.3 and 2022.3 all LTS versions
* Graphics API: SRP, URP, HDRP
* Browser: Chrome, Firefox, Edge, Safari ( all latest )
```

### Steps To Reproduce

1. Open bidirectional examples on the web and unity with different networks ( ask your friend for help )
2. Try to make a connection. 

### Current Behavior

It will look like there is a connection at least half of it but it will not detect the encoders and will not show up any data as the stream does not start. And disconnects in 60 seconds at max. If you try in the local network ( both web and unity apps ) it will connect yet get disconnected at an unmeasured time.

### Expected Behavior

It should connect and stay connected as long as the connection is closed by one of the peers.

### Anything else?

I have tried 3 different Turn server solutions with various configurations, with and without authentication.
All tests passed on ICE serves when I tested trickle ( except chrome with the known issues but it is considered passed )
I tried Coturn and Eturnal servers. Then I decided to try free and paid 3rd party servers ( metered.ca ). All passed the trickle and peer-to-peer tests.

I believe there is a race condition between negotiation and state changes. I couldn't see any rollback scenario so cable tangling may be the problem.

**Comment by karasusan** [comment]
OK, let me check the issue on my enviroment.

**Comment by karasusan** [comment]
memo: URS-436

**Comment by emrahtoy** [comment]
@karasusan is there any update? Could anyone able to reproduce the issue?

**Comment by karasusan** [comment]
@emrahtoy
I am checking the issue in this week. 
I will share update if I can reproduce it.

**Comment by streunerlein** [comment]
+1, we are fighting with this for weeks. It is pretty sure a negotiation problem. Would be great to have that resolved because somehow render streaming in the same network is ~nice~ but the point of WebRTC is kind of to cross networks.

**Comment by karasusan** [comment]
Hi, I fixed this issue #923.
Please try that.

**Comment by karasusan** [comment]
#923 is merged.

@emrahtoy 
Have you solved your issue?

**Comment by emrahtoy** [comment]
@karasusan we have moved to web side solutions which is why I couldn't have tested it yet. I will inform you as soon as I test and get the results.

**Comment by cr-zhichen** [comment]
Thank you very much for quickly fixing this issue. I have already made some deep modifications on the basis of Exp5. Once my modifications are stable, I will migrate to the new version for testing. If testing begins and results are available, I will inform you promptly.

**Comment by nathansegers** [comment]
I have tried the same things with the package of exp.7, and encounter the same issue.

I open the web client on 4G, when it's hosted on GCP.
I use the free tier TURN server as mentioned above by someone else.
The connection is setup from the Client side, as well as from Unity.

There does not seem to be an error, but the messages are stuck at Receiging / Sending offers, answers and candidates.
No negotiation seems to be finishing off.
Is there an easy way to test that Unity can connect to the TURN server?

I'll try with the samples too, to see if it's my issue or not.

**Comment by nathansegers** [comment]
Update:
Seems like I still can't connect the Unity application in some cases.
It seemed to work, but only in rare scenario's. I'm trying to pinpoint the exact problem here ...

**Comment by VanIseghemThomas** [comment]
Can confirm I have the same behaviour as @nathansegers. The streaming works sometimes across public network, but only once in like 20 times. Devices on the same network have no problem.

**Comment by aryansaurav** [comment]
same issue encountered! I also suspect something to do with the TURN/STUN servers.

@nathansegers would you mind sharing details of the TURN server you used? Are there any credentials and is it free to use?

**Comment by nathansegers** [comment]
@aryansaurav I am using the default configuration of `metered.ca`.
My frontend, and the sample from this plugin, is taking the config directly from the API.
In Unity, I have set the same configuration with username and password.

I can share you the details over mail, if you are OK with that. You can send me an email on nathan@spectralbot.com

**Comment by VanIseghemThomas** [comment]
Update: confirmed TURN and STUN servers are configured correctly and functional.

**Comment by soursoup90** [comment]
The solution provided in the fix does not seem to work for me. I have noted, however, that modifying the following block of code in `SignalingManager.cs` ([source](https://github.com/Unity-Technologies/UnityRenderStreaming/blob/3.1.0-exp.7/com.unity.renderstreaming/Runtime/Scripts/SignalingManager.cs#L208-L209)) as suggested in https://github.com/Unity-Technologies/UnityRenderStreaming/issues/502#issuecomment-1526610985 yields promising results.

From:

```csharp
RTCConfiguration _conf = conf.GetValueOrDefault(new RTCConfiguration { iceServers = iceServers });
```

To:

```csharp
RTCConfiguration _conf = new RTCConfiguration { iceServers = iceServers };
```

This change allows the successful exchange of ICE candidates, enabling the Unity application to send `relay` candidates. Consequently, the WebRTC session gets successfully initiated.

Without the above change, I was getting the following errors in the console:

```
NullReferenceException: Object reference not set to an instance of an object
UnityEditor.SceneView.CallOnPreSceneGUI () (at /Users/bokken/build/output/unity/unity/Editor/Mono/SceneView/SceneView.cs:3693)
UnityEditor.SceneView.DoOnPreSceneGUICallbacks (UnityEngine.Rect cameraRect) (at /Users/bokken/build/output/unity/unity/Editor/Mono/SceneView/SceneView.cs:2248)
UnityEditor.SceneView.DoOnGUI () (at /Users/bokken/build/output/unity/unity/Editor/Mono/SceneView/SceneView.cs:2330)
UnityEditor.SceneView.OnSceneGUI () (at /Users/bokken/build/output/unity/unity/Editor/Mono/SceneView/SceneView.cs:2255)
UnityEngine.UIElements.IMGUIContainer.DoOnGUI (UnityEngine.Event evt, UnityEngine.Matrix4x4 parentTransform, UnityEngine.Rect clippingRect, System.Boolean isComputingLayout, UnityEngine.Rect layout
```

```
GUI Error: You are pushing more GUIClips than you are popping. Make sure they are balanced.
UnityEngine.GUIUtility:ProcessEvent (int,intptr,bool&) (at /Users/bokken/build/output/unity/unity/Modules/IMGUI/GUIUtility.cs:189)
```

**Comment by soursoup90** [comment]
For some reason I had the `-batchmode` flag enabled in my setup. I observed correct ICE candidates being sent, but there was no video stream ("black" screen) since the Unity Render Streaming does not support the headless mode.

After disabling the flag, I get error:

```
ArgumentException: Could not instantiate RTCPeerConnection
  at Unity.WebRTC.RTCPeerConnection..ctor (Unity.WebRTC.RTCConfiguration& configuration) [0x00066] in <4388613e512f48448b3620f52187f987>:0
  at Unity.RenderStreaming.PeerConnection..ctor (System.Boolean polite, Unity.WebRTC.RTCConfiguration config, System.Single resendInterval, System.Func`2[T,TResult] startCoroutine, System.Action`1[T] stopCoroutine) [0x0002f] in <5ece0a5662694536b471a11b7ee9fe68>:0
  at Unity.RenderStreaming.SignalingManagerInternal.CreatePeerConnection (System.String connectionId, System.Boolean polite) [0x0002f] in <5ece0a5662694536b471a11b7ee9fe68>:0
  at Unity.RenderStreaming.SignalingManagerInternal.OnCreateConnection (Unity.RenderStreaming.Signaling.ISignaling signaling, System.String connectionId, System.Boolean polite) [0x00000] in <5ece0a5662694536b471a11b7ee9fe68>:0
  at Unity.RenderStreaming.Signaling.WebSocketSignaling+<>c__DisplayClass39_0.<WSProcessMessage>b__0 (System.Object d) [0x0000f] in <5ece0a5662694536b471a11b7ee9fe68>:0
  at UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () [0x00002] in <6328b00c1acf464b8e65941bc42c4872>:0
  at UnityEngine.UnitySynchronizationContext.Exec () [0x00056] in <6328b00c1acf464b8e65941bc42c4872>:0
  at UnityEngine.UnitySynchronizationContext.ExecuteTasks () [0x00014] in <6328b00c1acf464b8e65941bc42c4872>:0

**Comment by nathansegers** [comment]
> 

Hi there!

This seems to have fixed our problem as well.
I don't know why, but it is solved and I can call my colleagues from Unity now!

Thank you for helping with the fixes!

@karasusan Will you check this in the source code and fix it as well?

**Comment by VanIseghemThomas** [comment]
Can also confirm this solves the issue. Thanks @soursoup90!

**Comment by bzurkowski** [comment]
I investigated further and discovered that the session setup fails only if I enable the `Evaluate Commandline Arguments` option in the `Signaling Manager` panel, regardless of whether CLI arguments are provided to the Unity command. After disabling the flag, the streaming works fine. Hope this provides some hint on fixing the bug.

cc @karasusan

**Comment by soursoup90** [comment]
I investigated further and discovered that the session setup fails only if I enable the `Evaluate Commandline Arguments` option in the `Signaling Manager` panel, regardless of whether CLI arguments are provided to the Unity command. After disabling the flag, the streaming works fine. Hope this provides some hint on fixing the bug.

cc @karasusan

**Comment by karasusan** [comment]
@soursoup90
Hi, let me check.

**Comment by aryansaurav** [comment]
Glad to know that @soursoup90 solution worked for so many people but I am a bit lost here! For me, the problem was only about TURN server, which I resolved with @nathansegers help. And now it works perfectly from the desktop client, though from mobile it is really really slow.

I just can't understand what issue @soursoup90  is talking about and why the proposed solution is needed. I see in the screenshot a Main camera in the Signaling Handler List of the Signaling Manager but I don't think it's needed. 

In my configuration that works perfectly, I have checked "Use Default Settings" and in the RenderStreamingSample.asset (inside Example folder) I have checked Automatic Streaming. I don't get the same error messages as others here.

If somehow can help me understand what the issue under discussion is about, would appreciate it. May be my solution could help in that case? And, is anyone having trouble with the super slow speed from mobile device (even on 5G network).?

**Comment by nathansegers** [comment]
Hi @aryansaurav,

I think the reason thew code from @soursoup90 worked is not doing anything special for you, or not needed is because of the Default settings.

In my case, I can't use the automatic streaming, because the Connection ID is dynamic and can only be filled in after the remote side has created an ID for the peer-to-peer connection.
In your case, because you use the "Use Default Streaming" checkbox, the `conf.GetValueOrDefault` will work better, because you are using the default values.
See the code snippet:

`RTCConfiguration _conf = conf.GetValueOrDefault(new RTCConfiguration { iceServers = iceServers });`

It might simply be an issue of using the Default settings from the Project settings menu.

Regarding the handlers, it should contain atleast one handler in the SignalingManager, in order to achieve streaming as far as I know.

The streaming on my sides are almost in sync with what I achieved through Google Meet while testing with a colleague, so I don't experience much delay.

I'll gladly help you test if needed.

**Comment by aryansaurav** [comment]
Thanks @nathansegers for the detailed explanations. And glad your test working so well. I've one last question. Did you already try on android mobile/ iOS mobile or just desktop? Just curious if it works as well on the mobile platform as on desktop... in my case, the android mobile streams extremely slowly (even on Wifi, or 5G) at 720p resolution, which I don't think is related to the network quality but rather the speed of the codec.

**Comment by karasusan** [comment]
We released a new version 3.1.0-exp.7. Please let me know if you have any issues.
This ticket will be closed if no response by the end this week.

**Comment by GXY1996ssss** [comment]
> 我们发布了新版本 3.1.0-exp.7。如果您有任何问题，请告诉我。 如果在本周结束前没有回复，此票证将被关闭。

When I start the program with  -signalingUrl ws://127.0.0.1:8181, the following error occurs

ArgumentException: Could not instantiate RTCPeerConnection
  at Unity.WebRTC.RTCPeerConnection..ctor (Unity.WebRTC.RTCConfiguration& configuration) [0x0006d] in E:\UnityProject\CloudDemo\Library\PackageCache\com.unity.webrtc@3.0.0-pre.6\Runtime\Scripts\RTCPeerConnection.cs:473 
  at Unity.RenderStreaming.PeerConnection..ctor (System.Boolean polite, Unity.WebRTC.RTCConfiguration config, System.Single resendInterval, System.Func`2[T,TResult] startCoroutine, System.Action`1[T] stopCoroutine) [0x00038] in E:\UnityProject\CloudDemo\Packages\Local\com.unity.renderstreaming\Runtime\Scripts\PeerConnection.cs:78 
  at Unity.RenderStreaming.SignalingManagerInternal.CreatePeerConnection (System.String connectionId, System.Boolean polite) [0x00035] in E:\UnityProject\CloudDemo\Packages\Local\com.unity.renderstreaming\Runtime\Scripts\SignalingManagerInternal.cs:364
